sockeye/transformer.py

```diff
@@ -0,0 +1,298 @@
+                                                       depth_out=config.model_size,
+                                                       dropout=config.dropout_attention,
+                                                       prefix="%satt_enc_" % prefix)
+        self.residual_enc = TransformerResidual(num_hidden=config.model_size,
+                                                layer_normalization=config.layer_normalization,
+                                                dropout=config.dropout_residual,
+                                                prefix="%satt_enc_res" % prefix)
+        self.feed_forward = TransformerFeedForward(num_hidden=config.feed_forward_num_hidden,
+                                                   num_model=config.model_size,
+                                                   dropout=config.dropout_relu,
+                                                   prefix="%sff_" % prefix)
+        self.residual_ff = TransformerResidual(num_hidden=config.model_size,
+                                               layer_normalization=config.layer_normalization,
+                                               dropout=config.dropout_residual,
+                                               prefix="%sff_res" % prefix)
+
+    def __call__(self,
+                 target: mx.sym.Symbol,
+                 target_lengths: mx.sym.Symbol,
+                 target_max_length: int,
+                 target_bias: mx.sym.Symbol,
+                 source: mx.sym.Symbol,
+                 source_lengths: mx.sym.Symbol,
+                 source_max_length: int) -> mx.sym.Symbol:
+        target = self.residual_self(target,
+                                    self.self_attention(target, target_lengths,
+                                                        target_max_length, bias=target_bias),
+                                    target_max_length)
+        target = self.residual_enc(target,
+                                   self.enc_attention(target, target_max_length,
+                                                      source, source_lengths, source_max_length),
+                                   target_max_length)
+        target = self.residual_ff(target,
+                                  self.feed_forward(target, target_max_length),
+                                  target_max_length)
+        return target
+
+
+class TransformerResidual:
+    """
+    Residual connection with optional layer normalization.
+    """
+
+    def __init__(self,
+                 num_hidden: int,
+                 layer_normalization: bool,
+                 dropout: float,
+                 prefix: str) -> None:
+        self.num_hidden = num_hidden
+        self.dropout = dropout
+        self.layer_norm = None
+        self.prefix = prefix
+        if layer_normalization:
+            self.layer_norm = layers.LayerNormalization(num_hidden=self.num_hidden, prefix=self.prefix)
```

Comment:
I don't feel strongly about it. I looked a couple of times at the parameter dicts from the transformer and found this a bit more concise and shorter.